reading_comprehension/qanet_encoder.py

```diff
@@ -0,0 +1,345 @@
+                 num_convs: int,
+                 conv_kernel_size: int,
+                 num_attention_heads: int,
+                 use_positional_encoding: bool = True,
+                 dropout_prob: float = 0.1,
+                 layer_dropout_undecayed_prob: float = 0.1,
+                 attention_dropout_prob: float = 0) -> None:
+        super(QANetEncoderBlock, self).__init__()
+
+        assert input_dim == hidden_dim
+
+        self._use_positional_encoding = use_positional_encoding
+
+        self._conv_norm_layers: List[LayerNorm] = []
+        self._conv_layers: List[DepthwiseSeparableConv] = []
+
+        for conv_idx in range(num_convs):
+            conv_norm = LayerNorm(hidden_dim)
+            # call add_module because this module is not a member variable of this class
+            self.add_module(f"conv_layer_norm_{conv_idx}", conv_norm)
+            self._conv_norm_layers.append(conv_norm)
+            conv = DepthwiseSeparableConv(hidden_dim, hidden_dim, conv_kernel_size, activation="relu", dim=1)
+            self.add_module(f"conv_{conv_idx}", conv)
+            self._conv_layers.append(conv)
+
+        self.attn_norm_layer = LayerNorm(hidden_dim)
+        self.self_attn_layer = MemoryEfficientMultiHeadSelfAttention(num_heads=num_attention_heads,
+                                                                     input_dim=hidden_dim,
+                                                                     attention_dim=attention_projection_dim,
+                                                                     values_dim=attention_projection_dim,
+                                                                     attention_dropout_prob=attention_dropout_prob)
+        self.ffn_norm_layer = LayerNorm(hidden_dim)
+        self.ffn = FeedForward(hidden_dim,
+                               activations=[Activation.by_name('relu')(),
+                                            Activation.by_name('linear')()],
+                               hidden_dims=[feedforward_hidden_dim, hidden_dim],
+                               num_layers=2,
+                               dropout=dropout_prob)
+
+        self.dropout = Dropout(dropout_prob)
+        self.residual_with_layer_dropout = ResidualWithLayerDropout(layer_dropout_undecayed_prob)
+        self._input_dim = input_dim
+        self._output_dim = hidden_dim
+
+    @overrides
+    def get_input_dim(self) -> int:
+        return self._input_dim
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self._output_dim
+
+    @overrides
+    def is_bidirectional(self):
+        return False
+
+    @overrides
+    def forward(self, inputs: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:  # pylint: disable=arguments-differ
+
+        if self._use_positional_encoding:
+            output = add_positional_features(inputs)
+        else:
+            output = inputs
+
+        total_sublayers = len(self._conv_layers) + 2
+        sublayer_cnt = 0
```

Comment:
`count`, not `cnt` - avoid abbreviations.  I'd similarly recommend `attention` instead of `attn` and `feedforward` or `feed_forward` instead of `ffn`, but I don't feel quite as strongly about those.